org.eclipse.m2e.core/src/org/eclipse/m2e/core/internal/builder/MavenBuilderImpl.java

```diff
@@ -78,6 +83,8 @@ public class MavenBuilderImpl {
 
   private final List<IIncrementalBuildFramework> incrementalBuildFrameworks;
 
+  private final Map<IProject, ProjectBuildState> deltaState = new ConcurrentHashMap<>();
```

Comment:
Did you consider storing this as a session property on the project to get automatic cleanup of the map?

Reply:
@szarnekow sounds like a good idea, do you like to propose a PR? This is currently a bit experimental but seem to solve the issues we see with missing resource copy, so improvements are always welcome!

Reply:
I'd love to but I don't see a test for this so I'm not quite sure which problem this in fact solves and how to ensure that a refactoring doesn't break it.